File: Cpp/801 - 1000/994 Rotting Oranges.cpp
// 8ms, 100.0%; 10.1MB, 99.48%
static int x = []() {
	ios_base::sync_with_stdio(false);
	cin.tie(nullptr);
	return 0;
}();

class Solution
{
public:
	int orangesRotting(vector<vector<int>> &grid)
	{
		int row = grid.size(), col = grid[0].size(), fresh = 0, time = 0;
		stack<pair<int, int>> rotten;
		for(int i = 0; i < row; ++i) {
			for(int j = 0; j < col; ++j) {
				if(grid[i][j] == 1) {
					++fresh;
				} else if(grid[i][j] == 2) {
					rotten.emplace(i, j);
				}
			}
		}
		while(true) {
			stack<pair<int, int>> nxt;
			while(!rotten.empty()) {
				auto &p = rotten.top();
				int i = p.first, j = p.second;
				rotten.pop();
				if(i - 1 >= 0) helper(i - 1, j, fresh, grid, nxt);
				if(j - 1 >= 0) helper(i, j - 1, fresh, grid, nxt);
				if(i + 1 < row) helper(i + 1, j, fresh, grid, nxt);
				if(j + 1 < col) helper(i, j + 1, fresh, grid, nxt);
			}
			if(nxt.empty()) break;
			rotten.swap(nxt);
			++time;
		}
		return (fresh == 0) ? time : -1;
	}

private:
	void helper(int i, int j, int &f, vector<vector<int>> &grid, stack<pair<int, int>> &nxt)
	{
		if(grid[i][j] == 1) {
			--f, grid[i][j] = 2;
			nxt.emplace(i, j);
		}
	}
};

File: Python3/801 - 1000/994 Rotting Oranges.py
# 52ms, 60.92%; 13.1MB, 5.12%
class Solution:
    def orangesRotting(self, grid: list) -> int:
        time, row, col = 0, len(grid), len(grid[0])
        rotten = [(i, j) for i in range(row) for j in range(col) if grid[i][j] == 2]
        while True:
            nxt = []
            while rotten:
                i, j = rotten.pop()
                if i - 1 >= 0: self.__helper(i - 1, j, grid, nxt)
                if j - 1 >= 0: self.__helper(i, j - 1, grid, nxt)
                if i + 1 < row: self.__helper(i + 1, j, grid, nxt)
                if j + 1 < col: self.__helper(i, j + 1, grid, nxt)
            if not nxt: break
            rotten = nxt
            time += 1
        return -1 if sum([g.count(1) for g in grid]) else time

    def __helper(self, i, j, grid: list, nxt: list) -> None:
        if grid[i][j] == 1:
            grid[i][j] = 2
            nxt.append((i, j))
